docs/tutorials/tutorial_2.rst

```diff
@@ -0,0 +1,548 @@
+            U.push_back(u);
+            V.push_back(v);
+        }
+    }
+
+Now, let us move on to the creation test cases itself. The actual test cases for a problem involving test groups are challenging. The first step is to draw the Venn diagram of the subtasks.
+
+The properties of the subtasks of the above problem are :
+
+1. All test cases in subtask 1 and subtask 2 are also in subtask 3, and subtask 3 contains some other test cases that are neither in subtask 1 nor subtask 2.
+
+2. Some test cases are in both subtask 1 and subtask 2, and there are also test cases that is in subtask 1 but not in subtask 2, and there are also test cases that are in subtask 2 but not in subtask 2.
+
+Therefore, the Venn diagram looks like this
+
+.. image:: tutorial_2-venn_diagram.png
+
+We need to define a test group for each of the region in the Venn diagram. Therefore, we will have four test groups. Let us number it from 1 to 4 in the same order as the diagram above. Therefore, each test groups will have the following constraints in addition to the global constraints:
+
+1. M ≤ 20 and all W are equal
+2. M ≤ 20 and not all W are equal
+3. M > 20 and all W are equal
+4. M > 20 and not all W are equal
+
+For example, the test cases for the first test group can be something like this
+
+.. sourcecode:: cpp
+
+    void TestGroup1() {
+        Subtasks({1, 2, 3});
+
+        CASE(N = 2, M = 1, U = {0}, V = {1}, W = {1});
+        CASE(N = 21, M = 20, randomTree(N, U, V), W.assign(M, 1000));
+        CASE(N = 20, M = 20, randomGraph(N, M, U, V), W.assign(M, 1000));
+
+        for (int i = 0; i < 5; i++) {
+            CASE(N = rnd.nextInt(2, 21),
+                 M = rnd.nextInt(N - 1, 20),
+                 randomGraph(N, M, U, V),
+                 W.assign(M, rnd.nextInt(1, 1000)));
+        }
+    }
+
+It is a good practice to include the smallest case (M = 1) and the largest case (M = 20) satisfying the test group constraints. Since this test group also include subtask 2, we also need to make sure that all W are equal.
+
+The second test group can be something like this
+
+.. sourcecode:: cpp
+
+    void TestGroup2() {
+        Subtasks({1, 3});
+        
+        CASE(N = 4, M = 4,
+             U = {0, 1, 2, 0},
+             V = {1, 2, 0, 3},
+             W = {1, 1, 1, 2});
+        // We manually create a small test case where greedily choosing
+        // the first N - 1 edges with smallest weight will create a cycle.
+
+        CASE(N = 2, M = 2, U = {0, 1}, V = {1, 0}, W = {1, 2});
+        CASE(N = 21, M = 20, randomTree(N, U, V), randomWeight(M, W));
+
+        for (int i = 0; i < 5; i++) {
+            CASE(N = rnd.nextInt(2, 21),
+                 M = rnd.nextInt(N - 1, 20),
+                 randomGraph(N, M, U, V),
+                 randomWeight(M, W));
+        }
+    }
+
+Since this test group does not include subtask 2, W must not be equal for all elements. Therefore, the smallest case for this test group is M = 2.
```

Comment:
- is not included in subtask 2
- it must not be the case that W are equal for all elements